### PR TITLE
Custom query handler

### PR DESCRIPTION
**Pull request by maurolacy**
Adding a custom query handler for Osmosis queries.

Pool state query implemented.

~~Tests fail with "imports github.com/osmosis-labs/osmosis/v7/app: import cycle not allowed in test"~~. (fixed)

~~Publishing as draft for early review. Not sure about package names, folder structure, etc.~~

### CODE REVIEW

**Comment by ethanfrey** [comment]
Ah, nice fix. I was just hunting that down and didn't see you had fixed it

File: app/keepers.go [hunk 1]
```diff
@@ -123,6 +123,8 @@ import (
 	bech32ibckeeper "github.com/osmosis-labs/bech32-ibc/x/bech32ibc/keeper"
 	bech32ibctypes "github.com/osmosis-labs/bech32-ibc/x/bech32ibc/types"
 	bech32ics20keeper "github.com/osmosis-labs/bech32-ibc/x/bech32ics20/keeper"
+
+	owasm "github.com/osmosis-labs/osmosis/v7/app/wasm"
```
Comment (ethanfrey) [inline app/keepers.go:127]: why `osmosis/v7`?
I don't see this directory. Is this pinning to a commit?

File: app/wasm/keeper.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package wasm
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/osmosis-labs/osmosis/v7/app/wasm/types"
+	gammkeeper "github.com/osmosis-labs/osmosis/v7/x/gamm/keeper"
+)
+
+type Keeper struct {
```
Comment (ethanfrey) [inline app/wasm/keeper.go]: I wouldn't use the word keeper here. Querier or Plugin or QueryPlugin

File: app/wasm/query_plugin.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+		if contractQuery.PoolState != nil {
+			poolId := contractQuery.PoolState.PoolId
+
+			state, err := osmoKeeper.GetPoolState(ctx, poolId)
+			if err != nil {
+				return nil, sdkerrors.Wrap(err, "osmo pool state query")
+			}
+
+			var assets []wasmvmtypes.Coin
```
Comment (ethanfrey) [inline app/wasm/query_plugin.go]: I think there is a helper in x/wasm to convert  []sdk.Coin to []wasmvmtypes.Coin (and visa versa)

Comment (ethanfrey) [reply app/wasm/query_plugin.go]: Hmmm... I guess only the opposite directionL: https://github.com/CosmWasm/wasmd/blob/master/x/wasm/keeper/handler_plugin_encoders.go#L322-L333

Can you pull this out into a function, which we can upstream eventually?

File: app/wasm/wasm.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+import (
+	"github.com/CosmWasm/wasmd/x/wasm"
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+)
+
+func SetupWasmHandlers(
+	osmoKeeper ViewKeeper,
+) []wasmkeeper.Option {
+	queryPluginOpt := wasmkeeper.WithQueryPlugins(&wasmkeeper.QueryPlugins{
```
Comment (ethanfrey) [inline app/wasm/wasm.go:11]: Okay, I was going to suggest a refactor, but then I realized you will register custom message handlers here as well, so you need the array.

I would just rename to `RegisterCustomPlugins` or something

File: app/keepers.go [hunk 5]
```diff
@@ -368,6 +372,10 @@ func (app *OsmosisApp) InitNormalKeepers(
 	// The last arguments can contain custom message handlers, and custom query handlers,
 	// if we want to allow any custom callbacks
 	supportedFeatures := "iterator,staking,stargate"
+
+	app.OwasmKeeper = owasm.NewKeeper(app.GAMMKeeper)
+	wasmOpts = append(owasm.SetupWasmHandlers(app.OwasmKeeper), wasmOpts...)
```
Comment (ethanfrey) [inline app/keepers.go]: Nice one here.

File: app/keepers.go [hunk 6]
```diff
@@ -161,6 +163,8 @@ type appKeepers struct {
 	SuperfluidKeeper     *superfluidkeeper.Keeper
 	GovKeeper            *govkeeper.Keeper
 	WasmKeeper           *wasm.Keeper
+	// Osmosis wasm query interface
+	OwasmKeeper owasm.ViewKeeper
```
Comment (ethanfrey) [inline app/keepers.go]: I wonder if we need to add this here?

I would propose just making this a local variable (that is added to wasmOpts) and not include it here as it is not a keeper, just a conversion wrapper.

**Review by ethanfrey — COMMENTED** [review]
Nice start.

File: app/keepers.go (hunk 1 above)
Comment (maurolacy) [reply app/keepers.go:127]: This is what the IDE (Intellij) suggests. Thought it was correct. Doesn't this depends on `go.mod` or so? Will take a look at the other packages, to see what they import.

File: app/wasm/query_plugin.go (hunk 3 above)
Comment (maurolacy) [reply app/wasm/query_plugin.go]: I will.

File: app/keepers.go (hunk 5 above)
Comment (maurolacy) [reply app/keepers.go]: This comes from tgrade (as well as all the rest, that I more or less adapted here).

**Comment by maurolacy** [comment]
> Ah, nice fix. I was just hunting that down and didn't see you had fixed it

I was hesitant to push that, because I don't understand why it fixes the issue. But seems to be OK, so.

**Comment by ethanfrey** [comment]
> I was hesitant to push that, because I don't understand why it fixes the issue. But seems to be OK, so.

It's the "Go Way".

"Split into multiple packages to avoid circular dependencies."
Rather than "make a smarter compiler that can resolve most of these cases"

**Pull request merged**